### PR TITLE
Added AggregateAction interface and implementations for Combine and RemoveDuplicate Aggregate Actions

### LINKED ISSUE #844: Create interface for Aggregate Actions and create actions for RemoveDuplicates and Combine

### PR DESCRIPTION
**Pull request by graytaylor0**
### Description
* Added a new interface called AggregateAction. Users will create plugins that implement this interface to create custom actions for use in AggregateProcessor.
* Implemented an AggregateAction for removing duplicate Events
* Implemented an AggregateAction for combining Events 
 
### Issues Resolved
closes #844 
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateAction.java
Comment (dlvenable) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateAction.java:22]: AggregationAction authors are going to rely on the documentation in these Javadocs. So I'd like to elaborate on the groupState some. Here is one possible sentence, but I'm open to others than convey the important points.

> A map for AggregationActions to maintain user-defined for a group. The same map is provided for all events in a single group. This map is non-null.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java
Comment (dlvenable) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java:19]: I don't think we should include `closeWindow` yet. I think this approach is the right one so that we can support it, but in-case it isn't needed, let's not yet implement it.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java
Comment (dlvenable) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java:21]: Passing in Optional arguments is not a good practice.

Better to take in an `Event` and then implement `getEvent` to return `Optional.ofNullable(event)`.

Comment (dlvenable) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java:21]: The items in this response can change overtime, so a builder pattern will help here.

But, I'd prefer that we start by providing simple functions for common use-cases:

e.g. With no event:

```
AggregateActionResponse.emptyEventResponse()
```

Or with the Event:

```
AggregateActionResponse.fromEvent(event);
```

Then as the fields grows, we can add a full-on builder pattern instead.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java
Comment (dlvenable) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java:21]: The `groupState` should be guaranteed non-null by the `AggregateProcessor`. This check should be unnecessary.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java
Comment (dlvenable) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java:17]: Should this be a different event type? We may need to get some clarity on when to use different event types.

File: data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateActionTest.java
Comment (dlvenable) [inline data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateActionTest.java:78]: I recommend that you only test one function per test. If any assertions fail after the second call, it is difficult to tell which function was in error. This approach can also reduce the number of complexity of test cases.

The best solution will be to change the `handleEvent` method to make assertions on the `groupState`. Then in the test for `concludeGroup`, setup the state correctly.

File: data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateActionTest.java
Comment (dlvenable) [inline data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateActionTest.java:96]: In line with my comment above on testing one function, this can be simplified.

One test will test `handleEvent` adds an event to an empty `groupState`. A second test will start with a `groupState` with some events. Then it tests that `handleEvent` adds new events. These tests combined tell you recursively that further combinations will work.

File: data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateActionTest.java
Comment (dlvenable) [inline data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateActionTest.java:34]: I recommend that you use the `createObjectUnderTest()` pattern. This helps prevent calling methods in the setup method. It is more of a long-term maintainability benefit as the code changes hands over time.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java
Comment (graytaylor0) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java:17]: We haven't discussed different event types much, but I think it would be nice to have some ability to tell which Event is the final combined aggregate. I think it makes sense to have here.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java
Comment (graytaylor0) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java:21]: So for the case where the groupState doesn't exist for a certain group yet, will the `AggregateProcessor` just not call handleEvent?

Comment (graytaylor0) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java:21]: Or should this null check be replaced with something like `if (groupState.size() == 0)`

Comment (dlvenable) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java:21]: The AggregateProcessor should create a new `Map` when the groupState is created.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java
Comment (cmanning09) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java:17]: I am concerned we are going to clutter our event namespace. Let's remove it until we have a necessary use case for it.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateAction.java
Comment (cmanning09) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateAction.java:17]: Can you add documentation for how this AggregateAction works?

File: data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateActionTest.java
Comment (cmanning09) [inline data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateActionTest.java:22]: Are we missing the test which demonstrates receiving an event which is already in the group state does not update the group state?

Do I have a correct understanding of the remove_duplicates expected functionality?

**Review by cmanning09 — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java
Comment (cmanning09) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java:22]: The results of this aggregation seems more like an overwrite than a combination. Order of events received will impact the output. Is this the expected behavior?

File: data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateActionTest.java
Comment (graytaylor0) [reply data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateActionTest.java:22]: I can make a test for this, but the groupState is never modified directly by `remove_duplicates` even for the first one where the groupState doesn't contain anything. The logic of modifying the groupState for the first Event that is a part of that groupState will be handled by the `AggregateProcessor`. So the flow will be somewhat like this.

In AggregateProcessor:
* check if the groupState exists for a certain Event (if it doesn't then groupState is an empty `Map<Object, Object>)`
* Call `aggregateAction.handleEvent(groupState)`
* Add the starting Event to the `groupState`

With this logic RemoveDuplicates will never modify the groupState no matter what.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java
Comment (graytaylor0) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java:22]: I was a little confused on this too but @dlv discussed it. The idea is that there are the following logs which are grouped by the identification keys `sourceIp`, `destinationIp` and `port`. Sometimes different pieces of these logs will come in over time, so let's say you have this

` { sourceIp: 123, destinationIp: 456, port: 100, responseStatus: 200 } `

and then a little later you get 

` { sourceIp: 123, destinationIp: 456, port: 100, messageSize: 1,000 } `

then you want to combine these two together to have both messageSize and responseStatus

` { sourceIp: 123, destinationIp: 456, port: 100, messageSize: 1,000, responseStatus: 200 } `

I believe this scenario was framed to David by Cargill, so that is where the idea came from. Of course, we will get their feedback once we have some more implementation, and we can discuss with them if this is what they had in mind for their AggregateActions.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java
Comment (dinujoh) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java:18]: Optional was mainly intended for return values from methods, and not for properties of Java classes. We should consider replacing with with Event type. Also Optional is not serializable, so if we plan to serialize AggregateActionResponse object we will get unexpected data structure in the final output.

File: data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateActionTest.java
Comment (graytaylor0) [reply data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateActionTest.java:96]: My thought was that instead of making the AggregateAction creator check if groupState is empty and add the Event to it, I would handle this logic in the AggregateProcessor. Otherwise every single AggregateAction will have to copy and paste the following code.
```
if (groupState.size() == 0)
     addEventToGroupState
```

I suppose this is fine to have to do, but is it easier to not have to repeat this every time on the AggregateAction side? I guess it does give the AggregateAction creator the freedom to not add an Event to the groupState (even though this wouldn't really make sense to do)

File: data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateActionTest.java
Comment (graytaylor0) [reply data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateActionTest.java:22]: However, there is a consideration to let `AggregateAction` creators handle this logic. It would just be duplicated for pretty much all AggregateActions

File: data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateActionTest.java
Comment (dlvenable) [reply data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateActionTest.java]: I'm not suggesting changing the code. The current code should work whether it is the first event in a group or not.

But, tests are different. The tests should cover the important different cases. So what I'm suggesting is that we have:
1. A test case to represent the first event in a group. Yes, the `groupState` will be empty in the test setup.
2. A test case to represent adding an event to a group where some other event (or events) have already been added. This test will setup a `groupState` with data.

If both of these tests pass, then we can be reasonably certain that all sorts of combinations (two events, three events, etc.) will work.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateAction.java
Comment (dlvenable) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateAction.java:23]: I wouldn't say this:
> An empty groupState means that this is the first Event that will correspond with the given groupState.

It is up to the action whether this is true or not.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java
Comment (dlvenable) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java:33]: Now that this is `event`, I believe you can use `JacksonEvent.fromMessage`.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateAction.java
Comment (dlvenable) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateAction.java:25]: Rather than adding the whole event, I'd probably say just to add a special key. `groupState.put("atLeastOneEvent", true)`

This will more clearly indicate why you are putting any data.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java
Comment (dlvenable) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateAction.java:33]: Sorry, what I said above is not true.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java
Comment (dinujoh) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java:24]: Sorry for not being clear on my previous comment on the implication of having Optional in getter of POJO. If we try to serialize AggregateActionResponse object to JSON, we'll most likely see some unexpected data structure in the final JSON.

So, if we use Jackson to serialize AggregateActionResponse object, we'll get:
```
final ObjectMapper objectMapper = new ObjectMapper();
final Event event = constructEvent();
final AggregateActionResponse response = AggregateActionResponse.fromEvent(event);
System.out.println(objectMapper.writeValueAsString(response));
```

```
{"event":{"empty":false,"present":true}}
```

Consider returning Event for getEvent method and let the calling code convert this into Optional if needed.

```
public Event getEvent() { return event; }
```

Calling code

```
Optional<Event> event = Optional.ofNullable(aggregateActionResponse.getEvent());
```

Comment (graytaylor0) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java:24]: Is it an option to have 2 getters for the Event, one as the base Event and one as an Optional<Event>? I am pretty sure it is possible to tell Jackson which getter matters for serialization, and which one should be ignored.

Comment (dinujoh) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java:24]: yeah, you can use JsonProperty annotation to a method specify how to deserialize the attribute.

```
@JsonProperty("event")
public Event getEvent() { return event; }

public Event getOptionalEvent() { return Optional.ofNullable(event); }

```

The only problem with approach is we are coupled with Jackson serialization/deserialization. If we decide to use some other form of serialization, we need to make sure to change this.

Is there any reason you want to have method return optional ? Can the calling code have the wrapper around this method ? I would not prefer it for POJO or model class, but for non-pojo class it should be good.

Comment (graytaylor0) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java:24]: The reason I would prefer returning an Optional is to avoid returning null in the function. When the getter is used and Event is returned as null it makes NPEs much more likely. 

However, I did some reading and you are correct that it isn't encouraged to return Optionals is getters like this. I do think that the your code snippet above where the calling code wraps in an Optional is good. I will make sure to document in the javadoc for `AggregateActionResponse` that Event can be null.

Comment (dlvenable) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java]: I'd rather return only `Optional<Event>`. The field must be only `Event`. I believe Jackson will serialize this as an `Event` this way.

Comment (dlvenable) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java]: I'm also fine with `Event getEvent()` if that is what we agree on. Most importantly, I'd like only one getter.

**Review by dinujoh — APPROVED** [review]

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java
Comment (sbayer55) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java:16]: nit: Can event be final?

**Review by sbayer55 — APPROVED** [review]

**Review by dlvenable — APPROVED** [review]
I'm approving, but I would like to the `Event` field made private per @sbayer55 's comment.

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java
Comment (cmanning09) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionResponse.java:12]: Please update the documentation. This model does not contain "an option to close the current window"

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateAction.java
Comment (cmanning09) [inline data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateAction.java:28]: By returning a non null AggregateActionResponse are we indicating all duplicates have been removed?

**Review by cmanning09 — COMMENTED** [review]
I have some minor feedback but I think this all can be address in a future PR. No need to hold this work up.

**Review by cmanning09 — APPROVED** [review]

File: data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateAction.java
Comment (dlvenable) [reply data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/RemoveDuplicatesAggregateAction.java:28]: This will return the first Event. Then all future Events are suppressed.

**Comment by dlvenable** [comment]
> I'm approving, but I would like to the `Event` field made private per @sbayer55 's comment.

You have a whopping four approvals! I'm fine with fixing this in a follow-on PR.

**Pull request merged**
